README.md

```diff
@@ -5,33 +5,173 @@ state container for [Reason](https://github.com/facebook/reason) applications.
+
+`useSelector` is built on top of the [`useSubscription`](https://github.com/facebook/react/tree/master/packages/use-subscription) hook, which is [safe to use](https://github.com/facebook/react/tree/master/packages/use-subscription#limitations-in-concurrent-mode) in the concurrent mode.
+
+#### Selector function
+
+The selector function is required to have a **stable reference** in order to avoid infinite re-renders. The easiest to achieve this is to declare it outside of the component,
+
+```reason
+// declare selector outside of the component
+let userSelector = state => state.user;
+
+// in the component
+let user = AppStore.useSelector(selector);
+```
+
+or memoize with `useCallback` if it depends on `props`, `state` or anything else accessible only inside of the component:
+
+```reason
+[@react.component]
+let make = (~id) => {
+  let productSelector =
+    React.useCallback1(
+      state => state.products->Belt.List.keep(p => p.id === id),
+      [|id|],
+    );
+  let product = AppStore.useSelector(productSelector);
+  ...
+};
+```
+
+#### Re-renders
+
+`useSelector` relies on `useState` under the hood and therefore allows to bail-out of re-render similar to how `useState` [works](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-state-update), which will compare by value for primitive types, and by reference for objects.
+
+If the selector function returns an object, it won't cause a re-render only if the new object has the same **reference** as the previous one, and returning a new object every time will **always** cause a re-render. For example,
+
+```reason
+let selector = state => {email: state.user.email, cart: state.shop.cart};
+
+// in the component
+let selectedState = AppStore.useSelector(selector);
+```
+
+will cause the component to re-render every time an action is dispatched, regardless of whether `user` or `shop` have changed, since running the selector function will create a new object every time it is called. To prevent those re-renders, it is recommended to have multiple calls to `useSelector`, one per each individual field:
+
+```reason
+let emailSelector = state => state.user.email;
+let cartSelector = state => cart: state.shop.cart;
+
+// in the component
+let email = AppStore.useSelector(emailSelector);
+let cart = AppStore.useSelector(cartSelector);
+```
+
+This helps to make sure the component re-renders only when either `email` or `cart` on the store state changes.
+
+This is different from how `mapStateToProps`, if you are used to dealing with the traditional `redux` flow, since `mapStateToProps` will compare individual fields of the returned object.
+
+### useDispatch
+
+Returns the dispatch function from the store:
+
+```reason
+let dispatch = AppStore.useDispatch();
+...
+dispatch(Increment);
+```
+
+## Use case
```

Comment:
Thanks this is cool! In general I think we should still link to https://medium.com/@dan_abramov/you-might-not-need-redux-be46360cf367 and I think this section should be moved to the top. If we had a website I'd be fine with this order but since the readme is the docs I think people should be making an informed decision before digging in.

Reply:
Absolutely! I have added the link to the top section in the docs.